master/internal/api_checkpoint.go

```diff
@@ -73,17 +75,46 @@ func (m *Master) canDoActionOnCheckpoint(
 	return nil
 }
 
+func (m *Master) canDoActionOnCheckpointThroughModel(ctx context.Context, curUser model.User,
+	id uuid.UUID) error {
+	modelIDs, err := m.db.GetModelIDsAssociatedWithCheckpoint(id)
+	if err != nil {
+		return err
+	}
+	for _, id := range modelIDs {
+		model := &modelv1.Model{}
+		_ = m.db.QueryProto("get_model_by_id", model, id)
+		ok, _ := modelauth.AuthZProvider.Get().CanGetModel(ctx, curUser, model, *model.WorkspaceId)
+		if ok {
+			return nil
+		}
+	}
+
+	return status.Error(codes.PermissionDenied,
+		fmt.Sprintf("cannot access checkpoint: %s", id.String()))
+}
 func (a *apiServer) GetCheckpoint(
 	ctx context.Context, req *apiv1.GetCheckpointRequest,
 ) (*apiv1.GetCheckpointResponse, error) {
 	curUser, _, err := grpcutil.GetUser(ctx)
 	if err != nil {
 		return nil, err
 	}
-	if err = a.m.canDoActionOnCheckpoint(ctx, *curUser, req.CheckpointUuid,
-		expauth.AuthZProvider.Get().CanGetExperimentArtifacts); err != nil {
+	errE := a.m.canDoActionOnCheckpoint(ctx, *curUser, req.CheckpointUuid,
+		expauth.AuthZProvider.Get().CanGetExperimentArtifacts)
+	ckptUUID, err := uuid.Parse(req.CheckpointUuid)
+	if err != nil {
 		return nil, err
 	}
+	errM := a.m.canDoActionOnCheckpointThroughModel(ctx, *curUser, ckptUUID)
+
+	if errE != nil && errM != nil {
+		// allow viewing checkpoint through any model associated with checkpoint.
+		return nil,
+			errors.Errorf(
+				"checkpoint access error with experiment: %v, error with model(s): %v.",
+				errE.Error(), errM.Error())
+	}
```

Comment:
since this is an OR condition (have permission on experiment OR on model) can we do the 2nd check lazily? that is, only do the second model check `if errE != nil`, saving one or two unnecessary SQL queries.